{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MERFISH whole brain spatial transcriptomics (part 1)\n",
    "\n",
    "The MERFISH data consists of 59 coronal sections (10 micron thick) covering the whole anterior to posterior extent of the brain (~200 micron apart) from a single adult male mouse. Each section is assayed using the MERSCOPE platform with a 500 gene panel designed to enable us to map cells to the whole mouse brain taxonomy.\n",
    "\n",
    "The MERSCOPE platform uses combinatorial labeling, sequential imaging and error robust barcodes to obtain quantitative (counts) RNA-expression of 100s-1000s of genes. Each probe binds to a specific transcripts resulting in “dots” in the image. Cells are segmented using CellPose software based on DAPI/PolyT stains.\n",
    "\n",
    "Gene expression data is stored either as individual sections or all sections in a single file. The puprose of this set of notebooks is to provide and overview of the data, the file organization and how to combine the data and metadata through example use cases.\n",
    "\n",
    "You need to be connected to the internet to run this notebook and that you have downloaded the example data via the getting started notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata\n",
    "import time\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The prerequisite for running this notebook is that the data have been downloaded to local directory maintaining the organization from the manifest.json. **Change the download_base variable to where you have downloaded the data in your system.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "version = '20230830'\n",
    "download_base =  '/alzheimer/Roberto/Allen_Institute/abc_download_root'\n",
    "\n",
    "use_local_cache = False\n",
    "manifest_path = 'releases/%s/manifest.json' % version\n",
    "\n",
    "if not use_local_cache :\n",
    "    url = 'https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/' + manifest_path\n",
    "    manifest = json.loads(requests.get(url).text)\n",
    "else :\n",
    "    file = os.path.join(download_base,manifest_path)\n",
    "    with open(file,'rb') as f:\n",
    "        manifest = json.load(f)\n",
    "    \n",
    "metadata = manifest['file_listing']['MERFISH-C57BL6J-638850']['metadata']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "view_directory = os.path.join( download_base, \n",
    "                               manifest['directory_listing']['MERFISH-C57BL6J-638850']['directories']['metadata']['relative_path'], \n",
    "                              'views')\n",
    "cache_views = False\n",
    "if cache_views :\n",
    "    os.makedirs( view_directory, exist_ok=True )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data overview\n",
    "\n",
    "### Cell metadata\n",
    "\n",
    "Essential cell metadata is stored as a dataframe. Each row represents one cell indexed by a *cell label*.\n",
    "\n",
    "Each cell is associated with a *brain section label*, *donor label*, *donor genotype* and *donor sex*. Each cell also has a set of x, y, z coordinates generated by rotating each section so that it is upright with midline approximatedly in the middle of the frame. The z coordinate takes into account the phyiscal spacing between each section.\n",
    "\n",
    "Each cell is mapped to the whole mouse brain taxonomy using a hierachical correlation method, resulting in the assignment of a *cluster alias\" and *average correlation score*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cells =  3938808\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>brain_section_label</th>\n",
       "      <th>cluster_alias</th>\n",
       "      <th>average_correlation_score</th>\n",
       "      <th>feature_matrix_label</th>\n",
       "      <th>donor_label</th>\n",
       "      <th>donor_genotype</th>\n",
       "      <th>donor_sex</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1019171907102340387-1</th>\n",
       "      <td>C57BL6J-638850.37</td>\n",
       "      <td>1408</td>\n",
       "      <td>0.596276</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>7.226245</td>\n",
       "      <td>4.148963</td>\n",
       "      <td>6.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104095349101460194-1</th>\n",
       "      <td>C57BL6J-638850.26</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.641180</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>5.064889</td>\n",
       "      <td>7.309543</td>\n",
       "      <td>4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1017092617101450577</th>\n",
       "      <td>C57BL6J-638850.25</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.763531</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>5.792921</td>\n",
       "      <td>8.189973</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1018093344101130233</th>\n",
       "      <td>C57BL6J-638850.13</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.558073</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>3.195950</td>\n",
       "      <td>5.868655</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1019171912201610094</th>\n",
       "      <td>C57BL6J-638850.27</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.591009</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>5.635732</td>\n",
       "      <td>7.995842</td>\n",
       "      <td>4.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      brain_section_label  cluster_alias  \\\n",
       "cell_label                                                 \n",
       "1019171907102340387-1   C57BL6J-638850.37           1408   \n",
       "1104095349101460194-1   C57BL6J-638850.26           4218   \n",
       "1017092617101450577     C57BL6J-638850.25           4218   \n",
       "1018093344101130233     C57BL6J-638850.13           4218   \n",
       "1019171912201610094     C57BL6J-638850.27           4218   \n",
       "\n",
       "                       average_correlation_score feature_matrix_label  \\\n",
       "cell_label                                                              \n",
       "1019171907102340387-1                   0.596276       C57BL6J-638850   \n",
       "1104095349101460194-1                   0.641180       C57BL6J-638850   \n",
       "1017092617101450577                     0.763531       C57BL6J-638850   \n",
       "1018093344101130233                     0.558073       C57BL6J-638850   \n",
       "1019171912201610094                     0.591009       C57BL6J-638850   \n",
       "\n",
       "                          donor_label donor_genotype donor_sex         x  \\\n",
       "cell_label                                                                 \n",
       "1019171907102340387-1  C57BL6J-638850          wt/wt         M  7.226245   \n",
       "1104095349101460194-1  C57BL6J-638850          wt/wt         M  5.064889   \n",
       "1017092617101450577    C57BL6J-638850          wt/wt         M  5.792921   \n",
       "1018093344101130233    C57BL6J-638850          wt/wt         M  3.195950   \n",
       "1019171912201610094    C57BL6J-638850          wt/wt         M  5.635732   \n",
       "\n",
       "                              y    z  \n",
       "cell_label                            \n",
       "1019171907102340387-1  4.148963  6.6  \n",
       "1104095349101460194-1  7.309543  4.2  \n",
       "1017092617101450577    8.189973  4.0  \n",
       "1018093344101130233    5.868655  2.4  \n",
       "1019171912201610094    7.995842  4.4  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpath = metadata['cell_metadata']['files']['csv']['relative_path']\n",
    "file = os.path.join( download_base, rpath)\n",
    "cell = pd.read_csv(file, dtype={\"cell_label\":str})\n",
    "cell.set_index('cell_label',inplace=True)\n",
    "print(\"Number of cells = \", len(cell))\n",
    "cell.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We define a small helper fuction that uses pandas groupby function to see how many unique items are associated for each field and list them out if the number of items is small."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_column_info(df) :\n",
    "    \n",
    "    for c in df.columns :\n",
    "        grouped = df[[c]].groupby(c).count()\n",
    "        members = ''\n",
    "        if len(grouped) < 30 :\n",
    "            members = str(list(grouped.index))\n",
    "        print(\"Number of unique %s = %d %s\" % (c,len(grouped),members))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique brain_section_label = 59 \n",
      "Number of unique cluster_alias = 5320 \n",
      "Number of unique average_correlation_score = 3932069 \n",
      "Number of unique feature_matrix_label = 1 ['C57BL6J-638850']\n",
      "Number of unique donor_label = 1 ['C57BL6J-638850']\n",
      "Number of unique donor_genotype = 1 ['wt/wt']\n",
      "Number of unique donor_sex = 1 ['M']\n",
      "Number of unique x = 3938808 \n",
      "Number of unique y = 3938808 \n",
      "Number of unique z = 59 \n"
     ]
    }
   ],
   "source": [
    "print_column_info(cell)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also bring in the pivot table from the \"cluster annotation tutorial\" to associate each cell with terms at each cell type classification level and the corresponding color."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy_metadata = manifest['file_listing']['WMB-taxonomy']['metadata']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neurotransmitter</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>supertype</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_alias</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0082 L2 IT PPP-APr Glut_3</td>\n",
       "      <td>0326 L2 IT PPP-APr Glut_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0082 L2 IT PPP-APr Glut_3</td>\n",
       "      <td>0327 L2 IT PPP-APr Glut_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0081 L2 IT PPP-APr Glut_2</td>\n",
       "      <td>0322 L2 IT PPP-APr Glut_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0081 L2 IT PPP-APr Glut_2</td>\n",
       "      <td>0323 L2 IT PPP-APr Glut_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0081 L2 IT PPP-APr Glut_2</td>\n",
       "      <td>0325 L2 IT PPP-APr Glut_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              neurotransmitter          class                subclass  \\\n",
       "cluster_alias                                                           \n",
       "1                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "2                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "3                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "4                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "5                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "\n",
       "                               supertype                    cluster  \n",
       "cluster_alias                                                        \n",
       "1              0082 L2 IT PPP-APr Glut_3  0326 L2 IT PPP-APr Glut_3  \n",
       "2              0082 L2 IT PPP-APr Glut_3  0327 L2 IT PPP-APr Glut_3  \n",
       "3              0081 L2 IT PPP-APr Glut_2  0322 L2 IT PPP-APr Glut_2  \n",
       "4              0081 L2 IT PPP-APr Glut_2  0323 L2 IT PPP-APr Glut_2  \n",
       "5              0081 L2 IT PPP-APr Glut_2  0325 L2 IT PPP-APr Glut_2  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpath = taxonomy_metadata['cluster_to_cluster_annotation_membership_pivoted']['files']['csv']['relative_path']\n",
    "file = os.path.join( download_base, rpath)\n",
    "cluster_details = pd.read_csv(file,keep_default_na=False)\n",
    "cluster_details.set_index('cluster_alias', inplace=True)\n",
    "cluster_details.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neurotransmitter_color</th>\n",
       "      <th>class_color</th>\n",
       "      <th>subclass_color</th>\n",
       "      <th>supertype_color</th>\n",
       "      <th>cluster_color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_alias</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#266DFF</td>\n",
       "      <td>#64661F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#266DFF</td>\n",
       "      <td>#CCA73D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#002BCC</td>\n",
       "      <td>#99000D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#002BCC</td>\n",
       "      <td>#5C8899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#002BCC</td>\n",
       "      <td>#473D66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              neurotransmitter_color class_color subclass_color  \\\n",
       "cluster_alias                                                     \n",
       "1                            #2B93DF     #FA0087        #0F6632   \n",
       "2                            #2B93DF     #FA0087        #0F6632   \n",
       "3                            #2B93DF     #FA0087        #0F6632   \n",
       "4                            #2B93DF     #FA0087        #0F6632   \n",
       "5                            #2B93DF     #FA0087        #0F6632   \n",
       "\n",
       "              supertype_color cluster_color  \n",
       "cluster_alias                                \n",
       "1                     #266DFF       #64661F  \n",
       "2                     #266DFF       #CCA73D  \n",
       "3                     #002BCC       #99000D  \n",
       "4                     #002BCC       #5C8899  \n",
       "5                     #002BCC       #473D66  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpath = taxonomy_metadata['cluster_to_cluster_annotation_membership_color']['files']['csv']['relative_path']\n",
    "file = os.path.join( download_base, rpath)\n",
    "cluster_colors = pd.read_csv(file)\n",
    "cluster_colors.set_index('cluster_alias', inplace=True)\n",
    "cluster_colors.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>brain_section_label</th>\n",
       "      <th>cluster_alias</th>\n",
       "      <th>average_correlation_score</th>\n",
       "      <th>feature_matrix_label</th>\n",
       "      <th>donor_label</th>\n",
       "      <th>donor_genotype</th>\n",
       "      <th>donor_sex</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>neurotransmitter</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>supertype</th>\n",
       "      <th>cluster</th>\n",
       "      <th>neurotransmitter_color</th>\n",
       "      <th>class_color</th>\n",
       "      <th>subclass_color</th>\n",
       "      <th>supertype_color</th>\n",
       "      <th>cluster_color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1019171907102340387-1</th>\n",
       "      <td>C57BL6J-638850.37</td>\n",
       "      <td>1408</td>\n",
       "      <td>0.596276</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>7.226245</td>\n",
       "      <td>4.148963</td>\n",
       "      <td>6.6</td>\n",
       "      <td></td>\n",
       "      <td>04 DG-IMN Glut</td>\n",
       "      <td>038 DG-PIR Ex IMN</td>\n",
       "      <td>0141 DG-PIR Ex IMN_2</td>\n",
       "      <td>0515 DG-PIR Ex IMN_2</td>\n",
       "      <td>#666666</td>\n",
       "      <td>#16f2f2</td>\n",
       "      <td>#3D53CC</td>\n",
       "      <td>#CC7A3D</td>\n",
       "      <td>#73FFBF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104095349101460194-1</th>\n",
       "      <td>C57BL6J-638850.26</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.641180</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>5.064889</td>\n",
       "      <td>7.309543</td>\n",
       "      <td>4.2</td>\n",
       "      <td>Glut</td>\n",
       "      <td>23 P Glut</td>\n",
       "      <td>235 PG-TRN-LRN Fat2 Glut</td>\n",
       "      <td>0953 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>4199 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#6b5ca5</td>\n",
       "      <td>#9B7ACC</td>\n",
       "      <td>#990041</td>\n",
       "      <td>#663D63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1017092617101450577</th>\n",
       "      <td>C57BL6J-638850.25</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.763531</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>5.792921</td>\n",
       "      <td>8.189973</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Glut</td>\n",
       "      <td>23 P Glut</td>\n",
       "      <td>235 PG-TRN-LRN Fat2 Glut</td>\n",
       "      <td>0953 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>4199 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#6b5ca5</td>\n",
       "      <td>#9B7ACC</td>\n",
       "      <td>#990041</td>\n",
       "      <td>#663D63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1018093344101130233</th>\n",
       "      <td>C57BL6J-638850.13</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.558073</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>3.195950</td>\n",
       "      <td>5.868655</td>\n",
       "      <td>2.4</td>\n",
       "      <td>Glut</td>\n",
       "      <td>23 P Glut</td>\n",
       "      <td>235 PG-TRN-LRN Fat2 Glut</td>\n",
       "      <td>0953 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>4199 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#6b5ca5</td>\n",
       "      <td>#9B7ACC</td>\n",
       "      <td>#990041</td>\n",
       "      <td>#663D63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1019171912201610094</th>\n",
       "      <td>C57BL6J-638850.27</td>\n",
       "      <td>4218</td>\n",
       "      <td>0.591009</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>C57BL6J-638850</td>\n",
       "      <td>wt/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>5.635732</td>\n",
       "      <td>7.995842</td>\n",
       "      <td>4.4</td>\n",
       "      <td>Glut</td>\n",
       "      <td>23 P Glut</td>\n",
       "      <td>235 PG-TRN-LRN Fat2 Glut</td>\n",
       "      <td>0953 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>4199 PG-TRN-LRN Fat2 Glut_1</td>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#6b5ca5</td>\n",
       "      <td>#9B7ACC</td>\n",
       "      <td>#990041</td>\n",
       "      <td>#663D63</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      brain_section_label  cluster_alias  \\\n",
       "cell_label                                                 \n",
       "1019171907102340387-1   C57BL6J-638850.37           1408   \n",
       "1104095349101460194-1   C57BL6J-638850.26           4218   \n",
       "1017092617101450577     C57BL6J-638850.25           4218   \n",
       "1018093344101130233     C57BL6J-638850.13           4218   \n",
       "1019171912201610094     C57BL6J-638850.27           4218   \n",
       "\n",
       "                       average_correlation_score feature_matrix_label  \\\n",
       "cell_label                                                              \n",
       "1019171907102340387-1                   0.596276       C57BL6J-638850   \n",
       "1104095349101460194-1                   0.641180       C57BL6J-638850   \n",
       "1017092617101450577                     0.763531       C57BL6J-638850   \n",
       "1018093344101130233                     0.558073       C57BL6J-638850   \n",
       "1019171912201610094                     0.591009       C57BL6J-638850   \n",
       "\n",
       "                          donor_label donor_genotype donor_sex         x  \\\n",
       "cell_label                                                                 \n",
       "1019171907102340387-1  C57BL6J-638850          wt/wt         M  7.226245   \n",
       "1104095349101460194-1  C57BL6J-638850          wt/wt         M  5.064889   \n",
       "1017092617101450577    C57BL6J-638850          wt/wt         M  5.792921   \n",
       "1018093344101130233    C57BL6J-638850          wt/wt         M  3.195950   \n",
       "1019171912201610094    C57BL6J-638850          wt/wt         M  5.635732   \n",
       "\n",
       "                              y    z neurotransmitter           class  \\\n",
       "cell_label                                                              \n",
       "1019171907102340387-1  4.148963  6.6                   04 DG-IMN Glut   \n",
       "1104095349101460194-1  7.309543  4.2             Glut       23 P Glut   \n",
       "1017092617101450577    8.189973  4.0             Glut       23 P Glut   \n",
       "1018093344101130233    5.868655  2.4             Glut       23 P Glut   \n",
       "1019171912201610094    7.995842  4.4             Glut       23 P Glut   \n",
       "\n",
       "                                       subclass                    supertype  \\\n",
       "cell_label                                                                     \n",
       "1019171907102340387-1         038 DG-PIR Ex IMN         0141 DG-PIR Ex IMN_2   \n",
       "1104095349101460194-1  235 PG-TRN-LRN Fat2 Glut  0953 PG-TRN-LRN Fat2 Glut_1   \n",
       "1017092617101450577    235 PG-TRN-LRN Fat2 Glut  0953 PG-TRN-LRN Fat2 Glut_1   \n",
       "1018093344101130233    235 PG-TRN-LRN Fat2 Glut  0953 PG-TRN-LRN Fat2 Glut_1   \n",
       "1019171912201610094    235 PG-TRN-LRN Fat2 Glut  0953 PG-TRN-LRN Fat2 Glut_1   \n",
       "\n",
       "                                           cluster neurotransmitter_color  \\\n",
       "cell_label                                                                  \n",
       "1019171907102340387-1         0515 DG-PIR Ex IMN_2                #666666   \n",
       "1104095349101460194-1  4199 PG-TRN-LRN Fat2 Glut_1                #2B93DF   \n",
       "1017092617101450577    4199 PG-TRN-LRN Fat2 Glut_1                #2B93DF   \n",
       "1018093344101130233    4199 PG-TRN-LRN Fat2 Glut_1                #2B93DF   \n",
       "1019171912201610094    4199 PG-TRN-LRN Fat2 Glut_1                #2B93DF   \n",
       "\n",
       "                      class_color subclass_color supertype_color cluster_color  \n",
       "cell_label                                                                      \n",
       "1019171907102340387-1     #16f2f2        #3D53CC         #CC7A3D       #73FFBF  \n",
       "1104095349101460194-1     #6b5ca5        #9B7ACC         #990041       #663D63  \n",
       "1017092617101450577       #6b5ca5        #9B7ACC         #990041       #663D63  \n",
       "1018093344101130233       #6b5ca5        #9B7ACC         #990041       #663D63  \n",
       "1019171912201610094       #6b5ca5        #9B7ACC         #990041       #663D63  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_extended = cell.join(cluster_details,on='cluster_alias')\n",
    "cell_extended = cell_extended.join(cluster_colors,on='cluster_alias')\n",
    "cell_extended.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique brain_section_label = 59 \n",
      "Number of unique cluster_alias = 5320 \n",
      "Number of unique average_correlation_score = 3932069 \n",
      "Number of unique feature_matrix_label = 1 ['C57BL6J-638850']\n",
      "Number of unique donor_label = 1 ['C57BL6J-638850']\n",
      "Number of unique donor_genotype = 1 ['wt/wt']\n",
      "Number of unique donor_sex = 1 ['M']\n",
      "Number of unique x = 3938808 \n",
      "Number of unique y = 3938808 \n",
      "Number of unique z = 59 \n",
      "Number of unique neurotransmitter = 10 ['', 'Chol', 'Dopa', 'GABA', 'GABA-Glyc', 'Glut', 'Glut-GABA', 'Hist', 'Nora', 'Sero']\n",
      "Number of unique class = 34 \n",
      "Number of unique subclass = 338 \n",
      "Number of unique supertype = 1201 \n",
      "Number of unique cluster = 5320 \n",
      "Number of unique neurotransmitter_color = 10 ['#03EDFF', '#0a9964', '#2B93DF', '#533691', '#666666', '#73E785', '#820e57', '#FF3358', '#fcf04b', '#ff7621']\n",
      "Number of unique class_color = 34 \n",
      "Number of unique subclass_color = 338 \n",
      "Number of unique supertype_color = 1201 \n",
      "Number of unique cluster_color = 5296 \n"
     ]
    }
   ],
   "source": [
    "print_column_info(cell_extended)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For convenience, we can cache this view for later reuse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "if cache_views :\n",
    "    file = os.path.join( view_directory, 'cell_metadata_with_cluster_annotation.csv')\n",
    "    cell_extended.to_csv( file )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene panel\n",
    "\n",
    "A 500 gene panel was selected to enable faciliate the mapping to the whole mouse brain taxonomy. Each gene is uniquely identifier by an Ensembl ID. It is best practice to gene identifier to for tracking and data interchange as gene symbols are not unique and can change over time. The expression matrices also includes data for 50 \"blanks\" codewords in the spatial transcriptomic codebook that are not associated with any gene. They are included to measure the false discovery rate of gene detection.\n",
    "\n",
    "Each row of the gene dataframe has Ensembl gene identifier, the identifier of the transcript the probe targets, a gene symbol and name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes =  550\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>transcript_identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>mapped_ncbi_identifier</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000026778</th>\n",
       "      <td>Prkcq</td>\n",
       "      <td>ENSMUST00000028118</td>\n",
       "      <td>protein kinase C, theta</td>\n",
       "      <td>NCBIGene:18761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000026837</th>\n",
       "      <td>Col5a1</td>\n",
       "      <td>ENSMUST00000028280</td>\n",
       "      <td>collagen, type V, alpha 1</td>\n",
       "      <td>NCBIGene:12831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000001985</th>\n",
       "      <td>Grik3</td>\n",
       "      <td>ENSMUST00000030676</td>\n",
       "      <td>glutamate receptor, ionotropic, kainate 3</td>\n",
       "      <td>NCBIGene:14807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000039323</th>\n",
       "      <td>Igfbp2</td>\n",
       "      <td>ENSMUST00000047328</td>\n",
       "      <td>insulin-like growth factor binding protein 2</td>\n",
       "      <td>NCBIGene:16008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000048387</th>\n",
       "      <td>Osr1</td>\n",
       "      <td>ENSMUST00000057021</td>\n",
       "      <td>odd-skipped related transcription factor 1</td>\n",
       "      <td>NCBIGene:23967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_symbol transcript_identifier  \\\n",
       "gene_identifier                                        \n",
       "ENSMUSG00000026778       Prkcq    ENSMUST00000028118   \n",
       "ENSMUSG00000026837      Col5a1    ENSMUST00000028280   \n",
       "ENSMUSG00000001985       Grik3    ENSMUST00000030676   \n",
       "ENSMUSG00000039323      Igfbp2    ENSMUST00000047328   \n",
       "ENSMUSG00000048387        Osr1    ENSMUST00000057021   \n",
       "\n",
       "                                                            name  \\\n",
       "gene_identifier                                                    \n",
       "ENSMUSG00000026778                       protein kinase C, theta   \n",
       "ENSMUSG00000026837                     collagen, type V, alpha 1   \n",
       "ENSMUSG00000001985     glutamate receptor, ionotropic, kainate 3   \n",
       "ENSMUSG00000039323  insulin-like growth factor binding protein 2   \n",
       "ENSMUSG00000048387    odd-skipped related transcription factor 1   \n",
       "\n",
       "                   mapped_ncbi_identifier  \n",
       "gene_identifier                            \n",
       "ENSMUSG00000026778         NCBIGene:18761  \n",
       "ENSMUSG00000026837         NCBIGene:12831  \n",
       "ENSMUSG00000001985         NCBIGene:14807  \n",
       "ENSMUSG00000039323         NCBIGene:16008  \n",
       "ENSMUSG00000048387         NCBIGene:23967  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpath = metadata['gene']['files']['csv']['relative_path']\n",
    "file = os.path.join( download_base, rpath)\n",
    "gene = pd.read_csv(file)\n",
    "gene.set_index('gene_identifier',inplace=True)\n",
    "print(\"Number of genes = \", len(gene))\n",
    "gene.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene expression matrix\n",
    "\n",
    "Expression values for all sections are stored in an anndata h5ad format with minimal metadata. There are two h5ad files one storing the raw counts and the other log normalization of it. In this next section, we provide example code on how to open the file and connect with the rich cell level metadata discussed above.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_matrices = manifest['file_listing']['MERFISH-C57BL6J-638850']['expression_matrices']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'log2': {'files': {'h5ad': {'version': '20230830',\n",
       "    'relative_path': 'expression_matrices/MERFISH-C57BL6J-638850/20230830/C57BL6J-638850-log2.h5ad',\n",
       "    'url': 'https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/expression_matrices/MERFISH-C57BL6J-638850/20230830/C57BL6J-638850-log2.h5ad',\n",
       "    'size': 7627589574}}},\n",
       " 'raw': {'files': {'h5ad': {'version': '20230830',\n",
       "    'relative_path': 'expression_matrices/MERFISH-C57BL6J-638850/20230830/C57BL6J-638850-raw.h5ad',\n",
       "    'url': 'https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/expression_matrices/MERFISH-C57BL6J-638850/20230830/C57BL6J-638850-raw.h5ad',\n",
       "    'size': 7627589574}}}}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_matrices['C57BL6J-638850']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/alzheimer/Roberto/Allen_Institute/abc_download_root/expression_matrices/MERFISH-C57BL6J-638850/20230830/C57BL6J-638850-log2.h5ad\n"
     ]
    }
   ],
   "source": [
    "rpath = expression_matrices['C57BL6J-638850']['log2']['files']['h5ad']['relative_path']\n",
    "file = os.path.join( download_base, rpath)\n",
    "print(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use the anndata's *read_h5ad* function to open the package for the log2 normalization file. The \"backed='r'\" makes use of the lazy loading functionality to only load required data. By default, anndata will load the entire expression matrix in memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 4334174 × 550 backed at '/alzheimer/Roberto/Allen_Institute/abc_download_root/expression_matrices/MERFISH-C57BL6J-638850/20230830/C57BL6J-638850-log2.h5ad'\n",
      "    obs: 'brain_section_label'\n",
      "    var: 'gene_symbol', 'transcript_identifier'\n",
      "    uns: 'accessed_on', 'src'\n"
     ]
    }
   ],
   "source": [
    "adata = anndata.read_h5ad(file,backed='r')\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genes are represented as \"variables\". For this data, the var dataframe is indexed by the Ensembl gene identifier with one metadata column gene symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of genes = \", len(adata.var))\n",
    "adata.var.index[0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We define a small helper function *plot section* to visualize the cells in anatomical context colorized by: neurotransmitter identity, cell types division, class and subclass."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_section( xx, yy, cc=None, val=None, fig_width = 8, fig_height = 8, cmap=None ) :\n",
    "    fig, ax = plt.subplots()\n",
    "    fig.set_size_inches(fig_width, fig_height)\n",
    "    if cmap is not None :\n",
    "        plt.scatter(xx,yy,s=0.5,c=val,marker='.',cmap=cmap)\n",
    "    elif cc is not None :\n",
    "        plt.scatter(xx,yy,s=0.5,color=cc,marker='.')\n",
    "    ax.set_ylim(11,0)\n",
    "    ax.set_xlim(0,11)\n",
    "    ax.axis('equal')\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    return fig, ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let pick one section \"C57BL6J-638850.38\" for visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pred = (cell_extended['brain_section_label'] == 'C57BL6J-638850.38')\n",
    "section = cell_extended[pred]\n",
    "print(\"Number of cells in section:\",len(section))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plot_section(section['x'], section['y'], section['neurotransmitter_color'])\n",
    "res = ax.set_title(\"Neuortransmitter Identity\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plot_section(section['x'], section['y'],section['class_color'])\n",
    "res = ax.set_title(\"Cell Types: Class\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plot_section(section['x'], section['y'],section['subclass_color'])\n",
    "res = ax.set_title(\"Cell Types: Subclass\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example use cases\n",
    "\n",
    "In this section, we explore the same two use cases as with the 10x data. The first example looks at the expression of nine canonical neurotransmitter transporter genes and the second the expression of gene Tac2. To support these use cases, we will create a smaller submatrix (all cells and 10 genes) that read it into memory. *Note this operation takes around 10 seconds*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ntgenes = ['Slc17a7','Slc17a6','Slc17a8','Slc32a1','Slc6a5','Slc6a3','Slc6a4']\n",
    "exgenes = ['Tac2']\n",
    "gnames = ntgenes + exgenes\n",
    "pred = [x in gnames for x in adata.var.gene_symbol]\n",
    "gene_filtered = adata.var[pred]\n",
    "gene_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = time.process_time()\n",
    "asubset = adata[:,gene_filtered.index].to_memory()\n",
    "print(\"time taken: \", time.process_time() - start)\n",
    "print(asubset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We define two helper functions to (1) *create_expression_dataframe*: create joined gene expression and cell metadata dataframe based for a set of input genes and (2) *aggregate_by_metadata* which computes the average expression for each term in a given category."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_expression_dataframe( ad, gf ) :\n",
    "    gdata = ad[:,gf.index].to_df()\n",
    "    gdata.columns = gf.gene_symbol\n",
    "    joined = section.join( gdata )\n",
    "    return joined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aggregate_by_metadata( df, gnames, value, sort = False ) :\n",
    "    grouped = df.groupby(value)[gnames].mean()\n",
    "    if sort :\n",
    "        grouped = grouped.sort_values(by=gnames[0],ascending=False)\n",
    "    return grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression of canonical neurotransmitter transporter genes\n",
    "\n",
    "During analysis, clusters were assigned neurotransmitter identities based on the expression of of canonical neurotransmitter transporter genes. In this example, we create a dataframe comprising of cells in the 'C57BL6J-638850.40' section and expression of the 9 solute carrier family genes. We then group the cells by the assigned neurotransmitter class and compute the mean expression for each group and visualized as a colorized table. The results are similar of the 10x data restricted to cells from the thalamus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_heatmap( df, fig_width = 8, fig_height = 4, cmap=plt.cm.magma_r) :\n",
    "\n",
    "    arr = df.to_numpy()\n",
    "\n",
    "    fig, ax = plt.subplots()\n",
    "    fig.set_size_inches(fig_width,fig_height)\n",
    "\n",
    "    im = ax.imshow(arr, cmap = cmap, aspect='auto', vmin=0, vmax=5)\n",
    "    xlabs = df.columns.values\n",
    "    ylabs = df.index.values\n",
    "\n",
    "    ax.set_xticks(range(len(xlabs)))\n",
    "    ax.set_xticklabels( xlabs )\n",
    "\n",
    "    ax.set_yticks(range(len(ylabs)))\n",
    "    res = ax.set_yticklabels( ylabs )\n",
    "    \n",
    "    return im"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pred = [x in ntgenes for x in asubset.var.gene_symbol]\n",
    "gf = asubset.var[pred]\n",
    "\n",
    "ntexp = create_expression_dataframe( asubset, gf )\n",
    "agg = aggregate_by_metadata(ntexp, gf.gene_symbol, 'neurotransmitter')\n",
    "agg = agg[ntgenes]\n",
    "im = plot_heatmap(agg,8,3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The expression data can also be visualized as scatter point in anatomical context. Gene Slc17a7 is highly enriched in the cortex , hippocampus, medial habenula with some expression in the thalamus. On the other hand gene Slc32a1 is enriched in the striatum with sparse expression on the cortical plate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plot_section(ntexp['x'], ntexp['y'],val=ntexp['Slc17a7'],cmap=plt.cm.magma_r)\n",
    "res = ax.set_title('Gene Slc17a7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plot_section(ntexp['x'], ntexp['y'],val=ntexp['Slc32a1'],cmap=plt.cm.magma_r)\n",
    "res = ax.set_title('Gene Slc32a1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression of Tachykinin 2 (Tac2) in the whole brain\n",
    "\n",
    "In mice, the tachykinin 2 (Tac2) gene encodes neuropeptide called neurokinin B (NkB). Tac2 is produced by neurons in specific regions of the brain know to be invovled in emotion and social behavior. Based on [ISH data](https://mouse.brain-map.org/experiment/siv?id=77279001&imageId=77284584&initImage=ish&coordSystem=pixel&x=5384.5&y=3832.5&z=3) from the Allen Mouse Brain Atlas, Tac 2 is sparsely expressed in the mouse isocortex and densely enriched is specific subcortical regions such the medial habenula (MH), the amygdala and hypothalamus.\n",
    "\n",
    "In this example, we create a dataframe comprising of cells in section 'C57BL6J-638850.40' and expression values of Tac2 in those cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gf = asubset.var[asubset.var.gene_symbol == 'Tac2']\n",
    "tac2 = create_expression_dataframe( asubset, gf )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Grouping cells by neurotransmitter identites and computing the mean expression in each group, we can observed that Tac2 gene is highly enriched in cholinergic cell types with some expression in Glut-GABA types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agg = aggregate_by_metadata(tac2, gf.gene_symbol, 'neurotransmitter', True).head(8)\n",
    "im = plot_heatmap(agg,1,3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Grouping cells by cell types class, shows that Tac2 is enriched in class \"16 MH-LH Glut\" with cells restricted to the medial (MH) and lateral (LH) habenula and a mixture of glutamatergic and cholinergic types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agg = aggregate_by_metadata(tac2, gf.gene_symbol, 'class', True).head(8)\n",
    "im = plot_heatmap(agg,1,3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the next level, grouping by subclass, shows Tac2 is enriched in sublcasses with specificity to the medial habenula (MH), bed muclei of the stria terminalis (BST) and main olfactory blub (MOB)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agg = aggregate_by_metadata(tac2, gf.gene_symbol, 'subclass', True).head(8)\n",
    "im = plot_heatmap(agg,1,3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plotting Tac2 expression in anatomical coordinates results in a image that is comparable to the ISH image data showing enriched expression in the medial habenula, amydala and hypothalamus and sparse expression in the cortex."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plot_section(tac2['x'], tac2['y'],val=tac2['Tac2'],cmap=plt.cm.magma_r)\n",
    "res = ax.set_title('Gene Tac2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Close h5ad file and clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.file.close()\n",
    "del adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
