from Utils.Results_variables import *
from Utils.Settings import Adapt_for_Nature_style
from Utils.Utils import Naturize_text

results = {"Transcriptomic overview of 5-HT receptors landscape":
               "We analysed the single-cell RNA-sequencing (scRNA-seq) dataset provided by the Allen Institute {Yao, 2023 #2828} "
                "focusing on the expression of Htrs across approximately 4 million brain cells. The scRNA-seq dataset comprehensively encompassed all known 14 Htr subtypes. "
               f"Prevalence of Htrs across the entire dataset was considerably different ranging from {expression_total.min()}% of {expression_total.idxmin()} "
               f"to {expression_total.max()}% of {expression_total.idxmax()} (Figure 1A). RNA of 6 Htr was found in less than 2.5% of the cells "
               f"({', '.join(expression_total[expression_total < 2.5].index.values)}). On the other hand, RNA of Htr1f, Htr2a and Htr2c was present in "
               f"at least 1 every 5 cells. Average amount of RNA expression also varied across receptors (Supplementary Figure 1). "
               f"Interestingly, the variation in amount of RNA shared around half of the variability with the prevalence, i.e., receptors found in more cells also tended to be expressed more "
               f"on a single cell level. "
               f"Beside the prevalence and amount of expression, also the distribution among classes was considerably different. This is exemplified "
               f"by looking at the distribution of the Htr1 and Htr2 families across different cell type neighborhoods (Figure 1B). "
               f"Neighborhoods were defined both by location and neurotransmitter (Supplementary Figure 2). "
               f"When looking at the UMAP distributions for individual Htr, considerable differenceswere present also within family. We analyzed this differences more in detail by dividing cells by "
               f"neurotransmitter, neighborhoods and class. These categorizarions divided cells in a higlhy skewed manner (Supplementary Figure 3), "
               f"for example when looking at groups by neurotransmitter release, "
               f"3 groups (Glut, Gaba and unassigned) made up for almost the totality of cells. "
               f"Expectedly, the vast majority of cells was classified as excitatory ({total_cells_by_neurotransmitter['Glut']}%) and "
               f"around 1 every 5 cells was found to release GABA ({total_cells_by_neurotransmitter['GABA']}%). All the other neurotransmitter were found in less than 1% of the cells, "
               f"in particular, 5-HT releasing neurons (Sero) were found in only {total_cells_by_neurotransmitter['Sero']}% of the cells. "
               f"Some differences in Htrs distribution were visible across cells releasing different neurotransmitters (Figure 1C). However, the average "
               f"Pearson correlation coefficient (r) was surprisingly high at {round(expression_by_neurotransmitter.T.corr().mean().mean(),2)}±{round(expression_by_neurotransmitter.T.corr().mean().sem(),2)}."
               f"Sero and cholinergic neurons (Chol) showed the most distinct patterns of expression with respectively r={round(corr_by_neurotransmitter.mean()["Sero"],2)} and "
               f"r={round(corr_by_neurotransmitter.mean()["Chol"], 2)}. Cells not found to express any transmitter made out {total_cells_by_neurotransmitter['']}% of the total and "
               f"were the only group that did not express significant amounts of any Htrs (mean prevalence = {round(expression_by_neurotransmitter.T[''].mean(),2)}±{round(expression_by_neurotransmitter.T[''].sem(),2)}%). "
               f"All other groups expressed significant (defined as prevalence>20%) amounts of at least 2 different Htrs. "
               f"5-HT neurons exhibited the highest number of significant Htrs expression. They showed "
               f"the highest prevalence for all the receptors belonging to the Htr1 family (Htr1a: {expression_by_neurotransmitter['Htr1a']['Sero']}%, "
               f"Htr1b: {expression_by_neurotransmitter['Htr1b']['Sero']}%, Htr1d: {expression_by_neurotransmitter['Htr1d']['Sero']}% and Htr1f: {expression_by_neurotransmitter['Htr1f']['Sero']}%). "
               f"They also show significant amounts  of Htr2c ({expression_by_neurotransmitter['Htr2c']['Sero']}%), "
               f"Htr5a ({expression_by_neurotransmitter['Htr5a']['Sero']}%), Htr5b ({expression_by_neurotransmitter['Htr5b']['Sero']}%) "
               f"and Htr7 ({expression_by_neurotransmitter['Htr7']['Sero']}%). In total, Sero neurons showed significant expression of "
               f"{(expression_by_neurotransmitter>20).sum(axis=1)["Sero"]} different Htrs. Chol neurons distinguished themself by being particularly enriched in Htr4 and Htr5b."
               f"Looking at expression across neighborhoods, we noticed that non-neuronal cells (NN-IMN-GC) showed the lowest expression, mirroring the data regarding cells without "
               f"any neurotransmitter. Here the patterns of expression were even less differentiated across groups "
               f"(r={round(corr_by_group.mean(),2)}±{round(corr_by_group.sem(),2)}). "
               f"The totality of cells analyzed were divided in 34 classes following the original study. We analyzed expression across these  classes (Figure 1E). Average Pearson correlation between "
               f"patterns of expression was {round(corr_by_class.mean(), 2)}±{round(corr_by_class.sem(), 2)} (Figure 1F). Across classes, Htr2c was the one with the highest average prevalence "
               f"({mean_expression_by_class['Htr2c']}±{sem_expression_by_class['Htr2c']}%), followed by Htr1f ({mean_expression_by_class['Htr1f']}±{sem_expression_by_class['Htr1f']}%), "
               f"Htr7 ({mean_expression_by_class['Htr7']}±{sem_expression_by_class['Htr7']}%) and Htr4 ({mean_expression_by_class['Htr4']}±{sem_expression_by_class['Htr4']}%). "
               f"Correlation between Htrs expression across the totality of cells ranged from {corr_by_cell.min()} ({'-'.join(corr_by_cell.idxmin())}) to {corr_by_cell.max()} ({'-'.join(corr_by_cell.idxmax())}). "
               f"Considerable correlation was found also for the {'-'.join(corr_by_cell.index[1])} (Pearson coefficient={corr_by_cell.sort_values(ascending=False).iloc[1]}) and "
               f"{'-'.join(corr_by_cell.index[2])} (Pearson coefficient={corr_by_cell.sort_values(ascending=False).iloc[2]}) pairs (Figure 1H). Effect of this correlation was also visible "
               f"when looking at co-expression (Figure 1H). Expectedly, Htr1f and Htr2c, the most prevalent Htrs, were found to co-localize with other receptors respectevely {mean_coloc['Htr1f']}% "
               f"and  {mean_coloc['Htr2c']}% of the times. Only rarely a cell was found to express uniquely one Htr, {round(at_least_2_receptors.mean().values[0], 2)}±{round(at_least_2_receptors.sem().values[0], 2)}% "
               f"of cells indeed expressed at least "
               f"2 Htrs (Figure 1G).Surprisingly, {round(at_least_5_receptors.mean().values[0], 2)}±{round(at_least_5_receptors.sem().values[0], 2)}% of cells expressed at least 5 Htrs. This is indicative of the complexity "
               f"of the 5-HT system even at a single cell level. The highest amount of co-localization (at least 2 Htrs) was present in the GABAergic neurons of midbrain, hindbrain, and cerebellum (MB-HB-Glut-Sero-Dopa, "
               f"{round(at_least_2_receptors_per_group.loc['MB-HB-Glut-Sero-Dopa'].values[0], 2)}%). The average, excluding non neuronal cells, "
               f"was {round(at_least_2_receptors_per_group.drop('NN-IMN-GC').mean().values[0], 2)}±"
               f"{round(at_least_2_receptors_per_group.drop('NN-IMN-GC').sem().values[0], 2)}%. "
               f"The extensive expression across different classes and the considerable coexpression within cells point at the complexity of the 5-HT sistem. In the following sections we will take a deeper look "
               "each Htr family, also taking advantage of the spatial information provided by the MERFISH dataset of {Zhang, 2023 #2887} regarding 9 Htrs. "
               "This MERFISH dataset will be analyzed at different levels of granularity. "
               "In each of the following figures panels A, B and C are based on the RNA-seq dataset while the remaining on the MERFISH dataset.",
           "Htr1 & Htr5": u"Receptors belonging to these two families have an inhibitory effect on the host cell, they are coupled to G\u1D62 and cause a downstream decrease of cAMP "
                          "and activation of GIRK channels {Sharp, 2020 #2888; McCorvy, 2015 #2889}. "
                          "Hr1a RNA have a prevalence of around 10% in the brain, with a much lower presence in the TH-EPI-Glut and NN-IMN-GC groups (Figure 2A). The NN-IMN-GC group showed extremely "
                          "low expression across all receptors, for this reason it will be ignored.. "
                          "Htr1a co-localized most frequently with Htr1f, Htr2c and Htr2a (Figure 2B) and showed stable levels of co-localizations across groups. "
                          "Expression across classes was highly correlated between the RNA-seq and MERFISH datasets (Figure 2A) and show an almost perfect proportional relationship. "
                          "Highest expression was found in 5-HT neurons of the mid- and hindbrain (class 22 MB-HB Sero, Figure 2C). "
                          "Looking at the spatial distribution across divisions, the highest prevalence was found in cortical region of HPF, CTXsp and Isocortex (Figure 2D). At a more granular level, "
                          "the highest prevalence was observed in the medial septum (MS) and dorsal raphe (DR). DR expression is most likely effect of the high prevalence in 5-HT expressing neurons outlined above. "
                          "The cortical structures exhibiting higher prevalence were CA1 and medial entorhinal cortex (ENTm). "
                          "Expression was broadly distributed along the anterior-posterior axis (Figure 2E-F)"
                          "Htr1b exhibited a more diverse pattern of expression across groups (Figure 3A) with a range of 10 to 30%. Higher prevalence present in the MB-HB-Glut-Sero-Dopa group. "
                          "Prevalence in other groups, with the exception of NN-IMN-GC, was between 10 and 20% (Figure 3A).  Co-localization showed a similar pattern compared to Htr1a (Figure 3B)."
                          f"Looking at expression across classes, CNU-LGE GABA class showed the highest prevalence ({round(joined.groupby('class')['Htr1b'].apply(percentage_above_threshold).max(), 2)}%) "
                          f"in contrast with Htr1a that showed "
                          f"only minimal expression in this class ({round(joined.groupby('class')['Htr1a'].apply(percentage_above_threshold)['09 CNU-LGE GABA'], 2)}%). Prevalence > 20% was found also in a variety of other "
                          f"classes found in the hypothalamus (HY) and MB (Figure 3C). Striatum showed by far the highest prevalence with "
                          f"{merfish_by_gene ['Htr1b']['STR']} (Figure 3D-E-F). Caudoputamen (CP) and the lateral septal nucleus (LSc) exhibited a prevalence of almost 30%. "
                          f"Htr1d was expressed at a much lower level, never exceeding 8% prevalence across groups (Figure 4A). it colocalized at highest levels with "
                          f"Htr2c, Htr1f and Htr1b (Figure 4C). Similarly to Htr1b, expression was highest in 09 CNU-LGE GABA and 22 MB-HB Sero. "
                          f"Similarly, STR showed the highest expression (Figure 4D-E-F), however, the triangular nucleaus of septum (TRS) located in the pallidum (PAL) division, was "
                          f"the structure with highest prevalence. "
                          f"Interestingly, one cluster in the olfactory cortex (OLF), 0178 IT AON-TT-DP Glut, in the IT-ET Glut class exhibited high prevalence (>50%), "
                          f"visible on the left (black box) in Figure 4F. Average prevalence in the IT-ET Glut class was only "
                          f"{round(joined.groupby('class')['Htr1a'].apply(percentage_above_threshold)['01 IT-ET Glut'], 2)} %. "
                          f"Htr1f showed the highest expression of all 5-HT receptors in the RNA-seq dataset. Higher prevalence is found in the Pallium and Subpallium groups (Figure 5A), reaching almost 50%. "
                          f"Htr1f was found to co-localize the most with Htr2a and Htr2c (Figure 5B). Such high prevalence caused, however, lower levels of colocalization (Figure 5B). "
                          f"Notably, the slope of the linear regression between values provided by RNA-seq and MERFISH was "
                          f"significantly lower (Figure 5C). The two datasets are still highly correlated, half of the variability is shared. "
                          f"Spatial distribution showed a peculiarly asymettric pattern where expression was concentrated in the most anterior regions. Highest expression was observed in OLF, "
                          f"reaching over 20% consistently (Figure 5D-E-F). Specifically, highest expression was observed in the main (MOB) and accessory (AOB) olfactory bulbs. "
                          f"Both Htr5a and Htr5b were not included in the MERFISH dataset, therefore we do not have any spatial information regarding these two receptors. "
                          f"Htr5a was expressed at 10-15% prevalence across all groups with the exception of NN-IMN-GC (Supplementary Figure #) and colocalized the most with Htr1f, Htr2c and Htr2a. "
                          f"Prevalence a across did not show any clear peaks. "
                          f"Htr5b was expressed at a much lower level (Supplementary Figure #). Interestingly two classes accounted for the majority of the expression: 17 MH-LH Glut and 22 MB-HB Sero. ",

           "Htr2": "The Htr2 family is mainly linked to  Gq/11 and causes excitation by increasing intracellular Ca2+. Htr2a, famous for being instrumental in mediating the effects of psychedelics {Nichols, 2016 #854},"
                   " is found across the brain with highest prevalence in cortical groups, Pallium-Glut and Subpallium-GABA (Figure 6A). "
                   "Colocalization was highest with Htr1f and Htr2c (Figure 6B). Considerable expression (around 40%) was found in 01 IT-ET Glut, 07 CTX-MGE GABA and 16 HY-MM Glut classes (Figure 6C). "
                   "Similarly to Htr1f, also here the MERFISH dataset hinted at a lower overall expression when compared to RNA-seq. Shared variability between the two was, nonetheless very high. "
                   "Isocortex and CTXsp showed the highest prevalence, reaching more than 8% (Figure 6D). At a more detailed level, surprisingly, two structures belonging to the mammillary complex "
                   "(dorsal premammillary nucleus, PMd and tuberomammillary nucleus,TMd) "
                   "exhibited the highest prevalence. Claustrum (CLA) also showed high prevalence. Some subclasses in IT-ET Glut exhibited a particularly high prevalence, 001 CLA-EPd-CTX Car3 Glut and "
                          "027 L6b EPd Glut both had a prevalence of more than 90% (Supplementary Figure #). "
                   "Htr2b was found only in a minority of neurons and was not included in the MERFISH dataset. Interestigly, neurons belonging to "
                          f"the Pineal Glut class showed the highest prevalence at {round(joined.groupby('class')['Htr2b'].apply(percentage_above_threshold)['25 Pineal Glut'], 2)}% (Supplementary Figure #). "
                   f"Htr2c was found at highest prevalence in the MB-HB-Glut-Sero-Dopa and Hy-EA-Glut-Gaba groups (Figure 7A). Interestingly, groups with higher prevalence showed also higher levels of expression. "
                   f"Colocalization was highest with Htr1f, Htr4 and Htr7. In similar fashion to Htr2a, also here there were discrepancies between the RNA-seq and MERFISH methods (Figure 7C). "
                   f"Expression was broadly distributed across many different classes without a clear peak. The same applied to the spatial distribution, with no clear peaks (Figure 7D-E-F). "
                   f"HIghest prevalence was found in the MB, CTXsp and HY.",
           "Htr4, Htr6 and Htr7": "These receptor are all connected to Gs {McCorvy, 2015 #2889}, leading to increasing cellular levels of cAMP and excitation. "
                                  "Htr4, similarly to htr2C, showed highest prevalence (>40%) in the MB-HB-Glut-Sero-Dopa and Hy-EA-Glut-Gaba groups (Figure 8A). It colocalized the most with Htr2c and Htr1f (Figure 8B). "
                                  "Discrepancies in amount of expression between RNA-seq and MERFISH were present also here (Figure 8C). This did not affect notably, however, the correlation between the two datasets. "
                                  "Expression across classes did not exhibited any peculiar pattern. Spatial distribution,however, was more interesting, exhibiting a high prevalence in one specific structure of the STR, "
                                  "the olfactory tubercle (OT). "
                                  "We do not have spatial information about the rarely expressed Htr6. This receptor seemed to be expressed at significant prevalence only in the 09 CNU-LGE class () Supplementry Figure #. "
                                  "On the other habd, Htr7 was expressed in many more cells. It reached 60 % in the TH-EPI Glut group, and considerable amounts (around 40%) in MB, HB and HY groups (Figure 9A). "
                                  "Colocalization was the highest with Htr2c and Htr1f (Figure 9B). Expression was broadly distributed across classes present in HY, MB and TH (Figure 9C). This was reflected in the MERFISH dataset, "
                                  "showing highest prevalence in HY and TH (Figure 9D). At a structure level, the parafascicular nucleus of TH (PF) showed the highest prevalence (>40%). "
                                  "Notable expression was present in the most anterior parts of the brain (Figure 9E-F), areas belonging to OLF. "
                                  ""

           }

if Adapt_for_Nature_style is True:
    results = Naturize_text(results)
#%%
